### PR TITLE
LRU + Progressive throttling

### PR DESCRIPTION
**Pull request by asilvas-godaddy**
- **Feature** Mitigation strategy has shifted to use a lag range
  (between `minLag` & `maxLag`) which is used to determine at any given
  time how aggressive throttling should be. Throttling habits are now
  proportional to the lag/load, and throttling is prioritized based
  on the the worst offenders. This also removes the need for
  `waitForHistory`, `hostBadActorSplit`, and `ipBadActorSplit`.
  `minBadActorThreshold` & `maxBadActorThreshold` indicate the min/max
  range for the requests that will be blocked in proportion to the lag
- **Feature** Shifting strategies to an LRU in combination with
  `minHostRequests` & `minIpRequests` allows us to much more quickly
  begin blocking bad traffic (5x improvement at startup with default
  config), in addition to progressive updates as statistics are
  calculated in real time and no longer lag behind the giant
  `historySize` window to detect shifts in traffic patterns
- **Feature** With the addition of `exemptLocalAddress` we will no longer
  block (by default) or even track localhost requests, which is
  especially important for healthchecks not failing

### CODE REVIEW

File: src/connect.ts [hunk 1]
```diff
@@ -72,33 +127,57 @@ export class ConnectQOS {
...
-    if (this.tooBusy === true) {
-      const badActor = this.#metrics.isBadActor(req);
-      if (badActor) return badActor; // contains reason for bad actor
-      if (this.#userLag && this.lag >= this.#userLag) {
-        return BadActorType.userLag; // yes, throttle possibly innocent user
-      }
-    }
-  
+    // user lag takes priority even if no history is established to permit healthy recovery
+    if (this.lag >= this.#userLag) return BadActorType.userLag;
```
Comment (chinrichs-godaddy) [inline src/connect.ts]: I don't understand why userLag is checked first here - wouldn't that result in undercounting badHost/badIP traffic during attacks?

File: src/connect.ts [hunk 2]
```diff
@@ -72,33 +127,57 @@ export class ConnectQOS {
...
+    else if (this.isBadHost(req, false) === true) return BadActorType.badHost;
+    else if (this.isBadIp(req, false) === true) return BadActorType.badIp;
+
     // do not throttle user
     return false;  
   }
 
   get tooBusy() { // if metrics are not ready, we cannot be too busy
```
Comment (chinrichs-godaddy) [inline src/connect.ts]: Looks like this comment is no longer correct

File: src/connect.ts (hunk 1 above)
Comment (asilvas-godaddy) [reply src/connect.ts]: User lag has always taken priority, but for sake of reporting statistics I can see why this change might be undesirable. I'll revisit.

File: src/connect.ts [hunk 3]
```diff
@@ -72,33 +127,56 @@ export class ConnectQOS {
...
-      const badActor = this.#metrics.isBadActor(req);
-      if (badActor) return badActor; // contains reason for bad actor
-      if (this.#userLag && this.lag >= this.#userLag) {
-        return BadActorType.userLag; // yes, throttle possibly innocent user
-      }
-    }
-  
+    if (this.isBadHost(req, false) === true) return BadActorType.badHost;
+    else if (this.isBadIp(req, false) === true) return BadActorType.badIp;
+    else if (this.lag >= this.#userLag) return BadActorType.userLag;
```
Comment (chinrichs-godaddy) [inline src/connect.ts:149]: Should whitelists prevent requests from being blocked due to user lag? I would think probably yes, right?

Comment (asilvas-godaddy) [reply src/connect.ts:149]: Good catch. I'll decouple the whitelist from bad actor detection to clean this up.

**Review by chinrichs-godaddy — APPROVED** [review]
Nice updates!

**Pull request merged**